common/src/main/java/io/netty/util/internal/Cleaner.java

```diff
@@ -21,9 +21,21 @@
  * Allows to free direct {@link ByteBuffer}s.
  */
 interface Cleaner {
+    /**
+     * Create a direct {@link ByteBuffer} and return it alongside its cleaning mechanism,
+     * in a {@link CleanableDirectBuffer}.
+     *
+     * @param capacity The desired capacity of the direct buffer.
+     * @return The new {@link CleanableDirectBuffer} instance.
+     */
+    CleanableDirectBuffer allocate(int capacity);
 
     /**
      * Free a direct {@link ByteBuffer} if possible
+     *
+     * @deprecated Instead allocate buffers from {@link #allocate(int)}
+     * and use the associated {@link CleanableDirectBuffer#clean()} method.
      */
+    @Deprecated
     void freeDirectBuffer(ByteBuffer buffer);
```

Comment:
as this interface is package private we can also just directly remove it

Reply:
This still gets used via `Unpooled.wrapped(ByteBuffer)`. Maybe it shouldn't be deprecated.

Reply:
How is this be used ? It is package-private so can't be called directly ?

Reply:
If you create a `ByteBuf` with `Unpooled.wrapped(ByteBuffer)`, then we don't have a `CleanableDirectBuffer` instance. In that case, when releasing the `ByteBuf`, we go through this old code path.